File: ios/Classes/FlutterRecordSoundPlugin.swift
import Flutter
import UIKit
import AVFoundation

/**
 A Flutter plugin for recording audio on iOS.
 This plugin provides methods to start, stop, pause, resume recording, check permissions, and retrieve amplitude information.
 */
public class FlutterRecordSoundPlugin: NSObject, FlutterPlugin, AVAudioRecorderDelegate {
    
    // MARK: - Plugin Registration
    
    /**
     Registers the plugin with the Flutter engine.
     
     - Parameters:
        - registrar: The registrar provided by Flutter for registering plugins.
     */
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_record_sound", binaryMessenger: registrar.messenger())
        let instance = FlutterRecordSoundPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    // MARK: - Properties
    
    /// Indicates whether the recorder is currently recording.
    private var isRecording = false
    
    /// Indicates whether the recorder is currently paused.
    private var isPaused = false
    
    /// Indicates whether the app has permission to record audio.
    private var hasPermission = false
    
    /// The audio recorder instance used for recording.
    private var audioRecorder: AVAudioRecorder?
    
    /// The file path where the audio recording is saved.
    private var path: String?
    
    /// The maximum amplitude recorded so far.
    private var maxAmplitude: Float = -160.0
    
    /// A thread-safe queue for managing state variables.
    private let stateQueue = DispatchQueue(label: "com.flappy.flutter_record_sound.state")
    
    /// Thread-safe accessor for `isRecording`.
    private var isRecordingSafe: Bool {
        get { stateQueue.sync { isRecording } }
        set { stateQueue.sync { isRecording = newValue } }
    }
    
    /// Thread-safe accessor for `isPaused`.
    private var isPausedSafe: Bool {
        get { stateQueue.sync { isPaused } }
        set { stateQueue.sync { isPaused = newValue } }
    }
    
    // MARK: - Flutter Method Call Handling
    
    /**
     Handles method calls from Flutter.
     
     - Parameters:
        - call: The method call from Flutter.
        - result: The result callback to send data back to Flutter.
     */
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            let args = call.arguments as! [String: Any]
            path = args["path"] as? String
            if path == nil {
                // Generate a temporary file path if none is provided
                let directory = NSTemporaryDirectory()
                let fileName = UUID().uuidString + getFileExtension(for: args["encoder"] as? Int ?? 0)
                if let validPath = NSURL.fileURL(withPathComponents: [directory, fileName])?.absoluteString {
                    path = validPath
                } else {
                    result(FlutterError(code: "-2", message: "Failed to generate file path", details: nil))
                    return
                }
            }
            start(
                path: path!,
                encoder: args["encoder"] as? Int ?? 0,
                bitRate: args["bitRate"] as? Int ?? 128000,
                samplingRate: args["samplingRate"] as? Float ?? 44100.0,
                result: result
            )
        case "stop":
            stop(result)
        case "pause":
            pause(result)
        case "resume":
            resume(result)
        case "isPaused":
            result(isPausedSafe)
        case "isRecording":
            result(isRecordingSafe)
        case "hasPermission":
            hasPermission(result)
        case "getAmplitude":
            getAmplitude(result)
        case "dispose":
            dispose(result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - App Lifecycle Handling
    
    /**
     Stops recording when the app is about to terminate.
     */
    public func applicationWillTerminate(_ application: UIApplication) {
        stopRecording()
    }
    
    /**
     Stops recording when the app enters the background.
     */
    public func applicationDidEnterBackground(_ application: UIApplication) {
        stopRecording()
    }
    
    // MARK: - Permission Handling
    
    /**
     Checks and requests permission to record audio.
     
     - Parameters:
        - result: The result callback to send the permission status back to Flutter.
     */
    fileprivate func hasPermission(_ result: @escaping FlutterResult) {
        switch AVAudioSession.sharedInstance().recordPermission {
        case .granted:
            hasPermission = true
            result(hasPermission)
        case .denied:
            hasPermission = false
            result(hasPermission)
        case .undetermined:
            // Request permission if not determined
            AVAudioSession.sharedInstance().requestRecordPermission { [unowned self] allowed in
                DispatchQueue.main.async {
                    self.hasPermission = allowed
                    result(self.hasPermission) // Return result after permission request
                }
            }
        default:
            result(false)
        }
    }
    
    // MARK: - Recording Methods
    
    /**
     Starts recording audio with the specified parameters.
     
     - Parameters:
        - path: The file path where the recording will be saved.
        - encoder: The audio encoder type (e.g., AAC, AMR, WAV).
        - bitRate: The bit rate for the recording.
        - samplingRate: The sampling rate for the recording.
        - result: The result callback to notify Flutter about the recording status.
     */
    fileprivate func start(path: String, encoder: Int, bitRate: Int, samplingRate: Float, result: @escaping FlutterResult) {
        stopRecording() // Stop any ongoing recording
        
        let settings: [String: Any] = [
            AVFormatIDKey: getEncoder(encoder),
            AVEncoderBitRateKey: bitRate,
            AVSampleRateKey: samplingRate,
            AVNumberOfChannelsKey: 2,
            AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue,
            AVLinearPCMBitDepthKey: 16, // 16-bit PCM for WAV
            AVLinearPCMIsBigEndianKey: false,
            AVLinearPCMIsFloatKey: false
        ]
        
        let options: AVAudioSession.CategoryOptions = [.defaultToSpeaker, .allowBluetooth]
        
        DispatchQueue.global(qos: .background).async {
            do {
                // Configure the audio session
                try AVAudioSession.sharedInstance().setCategory(.playAndRecord, options: options)
                try AVAudioSession.sharedInstance().setActive(true)
                
                // Initialize the audio recorder
                let url = URL(string: path) ?? URL(fileURLWithPath: path)
                self.audioRecorder = try AVAudioRecorder(url: url, settings: settings)
                self.audioRecorder?.delegate = self
                self.audioRecorder?.isMeteringEnabled = true
                self.audioRecorder?.record()
                
                // Update state
                self.isRecordingSafe = true
                self.isPausedSafe = false
                
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "-1", message: "Failed to start recording", details: nil))
                }
            }
        }
    }
    
    /**
     Stops the recording and returns the file path.
     
     - Parameters:
        - result: The result callback to send the file path back to Flutter.
     */
    fileprivate func stop(_ result: @escaping FlutterResult) {
        stopRecording()
        result(path)
    }
    
    /**
     Pauses the recording.
     
     - Parameters:
        - result: The result callback to notify Flutter about the pause status.
     */
    fileprivate func pause(_ result: @escaping FlutterResult) {
        audioRecorder?.pause()
        isPausedSafe = true
        result(nil)
    }
    
    /**
     Resumes the recording.
     
     - Parameters:
        - result: The result callback to notify Flutter about the resume status.
     */
    fileprivate func resume(_ result: @escaping FlutterResult) {
        if isPausedSafe {
            audioRecorder?.record()
            isPausedSafe = false
        }
        result(nil)
    }
    
    /**
     Retrieves the current and maximum amplitude.
     
     - Parameters:
        - result: The result callback to send amplitude data back to Flutter.
     */
    fileprivate func getAmplitude(_ result: @escaping FlutterResult) {
        var amp: [String: Float] = ["current": -160.0]
        
        if isRecordingSafe {
            audioRecorder?.updateMeters()
            if let current = audioRecorder?.averagePower(forChannel: 0), current != -Float.infinity {
                if current > maxAmplitude {
                    maxAmplitude = current
                }
                amp["current"] = current
                amp["max"] = maxAmplitude
            }
        }
        
        result(amp)
    }
    
    /**
     Stops the recording and resets the state.
     */
    fileprivate func stopRecording() {
        if let recorder = audioRecorder, recorder.isRecording {
            recorder.stop()
        }
        audioRecorder = nil
        isRecordingSafe = false
        isPausedSafe = false
        maxAmplitude = -160.0
    }
    
    /**
     Releases resources and stops the recording.
     
     - Parameters:
        - result: The result callback to notify Flutter about the disposal status.
     */
    fileprivate func dispose(_ result: @escaping FlutterResult) {
        stopRecording()
        result(path)
    }
    
    // MARK: - Helper Methods
    
    /**
     Returns the encoder type for the given encoder ID.
     
     - Parameters:
        - encoder: The encoder ID.
     - Returns: The corresponding encoder type.
     */
    fileprivate func getEncoder(_ encoder: Int) -> Int {
        switch encoder {
        case 1:
            return Int(kAudioFormatMPEG4AAC_ELD)
        case 2:
            return Int(kAudioFormatMPEG4AAC_HE)
        case 3:
            return Int(kAudioFormatAMR)
        case 4:
            return Int(kAudioFormatAMR_WB)
        case 5:
            if #available(iOS 14.0, *) {
                return Int(kAudioFormatOpus)
            } else {
                return Int(kAudioFormatMPEG4AAC) // Fallback to AAC
            }
        case 6: // WAV format
            return Int(kAudioFormatLinearPCM)
        default:
            return Int(kAudioFormatMPEG4AAC)
        }
    }
    
    /**
     Returns the file extension for the given encoder ID.
     
     - Parameters:
        - encoder: The encoder ID.
     - Returns: The corresponding file extension.
     */
    fileprivate func getFileExtension(for encoder: Int) -> String {
        switch encoder {
        case 6:
            return ".wav" // WAV format
        case 3, 4:
            return ".amr" // AMR format
        case 5:
            return ".opus" // Opus format
        default:
            return ".m4a" // Default to AAC
        }
    }
}

File: android/src/main/kotlin/com/flappy/flutter_record_sound/FlutterRecordSoundPlugin.kt
package com.flappy.flutter_record_sound

import io.flutter.plugin.common.PluginRegistry.RequestPermissionsResultListener
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodCall
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import android.content.Context
import android.app.Activity
import java.io.IOException
import android.Manifest
import android.os.Build
import java.io.File

/** FlutterRecordSoundPlugin */
class FlutterRecordSoundPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware, RequestPermissionsResultListener {

    // Activity plugin binding for managing activity lifecycle
    private var activityPluginBinding: ActivityPluginBinding? = null

    // Pending result for permission requests
    private var pendingPermResult: MethodChannel.Result? = null

    // Method channel for communication between Flutter and native code
    private lateinit var channel: MethodChannel

    // Application context
    private var context: Context? = null

    // Current activity
    private var activity: Activity? = null

    // Recorder instance for managing audio recording
    private var recorder: Recorder? = null

    companion object {
        private const val RECORD_AUDIO_REQUEST_CODE = 1001 // Request code for audio recording permission
    }

    // Called when the plugin is attached to the Flutter engine
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        this.context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_record_sound")
        channel.setMethodCallHandler(this)
    }

    // Called when the plugin is detached from the Flutter engine
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
    }

    // Called when the plugin is attached to an activity
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        addBinding(binding)
    }

    // Called when the activity is detached for configuration changes
    override fun onDetachedFromActivityForConfigChanges() {
        removeBinding()
    }

    // Called when the activity is reattached after configuration changes
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        addBinding(binding)
    }

    // Called when the plugin is detached from an activity
    override fun onDetachedFromActivity() {
        removeBinding()
    }

    // Adds the activity binding and initializes the recorder
    private fun addBinding(binding: ActivityPluginBinding) {
        activityPluginBinding?.removeRequestPermissionsResultListener(this)
        activity = binding.activity
        activityPluginBinding = binding
        activityPluginBinding!!.addRequestPermissionsResultListener(this)
        this.recorder = Recorder(binding.activity)
    }

    // Removes the activity binding and cleans up resources
    private fun removeBinding() {
        activityPluginBinding?.removeRequestPermissionsResultListener(this)
        activity = null
        activityPluginBinding = null
        recorder?.close()
        recorder = null
        pendingPermResult = null
    }

    // Handles method calls from Flutter
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> handleStart(call, result)
            "stop" -> handleStop(result)
            "pause" -> handlePause(result)
            "resume" -> handleResume(result)
            "isPaused" -> handleIsPaused(result)
            "isRecording" -> handleIsRecording(result)
            "hasPermission" -> handleHasPermission(result)
            "getAmplitude" -> handleGetAmplitude(result)
            "dispose" -> handleDispose(result)
            else -> result.notImplemented()
        }
    }

    // Handles the "start" method call to start recording
    private fun handleStart(call: MethodCall, result: MethodChannel.Result) {
        val encoder = call.argument<Int>("encoder")
        val bitRate = call.argument<Int>("bitRate")
        val samplingRate = call.argument<Double>("samplingRate")
        var path = call.argument<String>("path")
        if (encoder == null || bitRate == null || samplingRate == null) {
            result.error("-4", "Invalid arguments for start method", null)
            return
        }
        if (path == null) {
            val outputDir = activity?.cacheDir
            if (outputDir == null) {
                result.error("-3", "Failed to access cache directory", null)
                return
            }
            try {
                val outputFile = File.createTempFile("audio", getFileExtension(encoder), outputDir)
                path = outputFile.path
            } catch (e: IOException) {
                result.error("-3", "Failed to create temporary file", e.message)
                return
            }
        }
        recorder?.start(path!!, encoder, bitRate, samplingRate, result)
    }

    // Handles the "stop" method call to stop recording
    private fun handleStop(result: MethodChannel.Result) {
        recorder?.stop(result) ?: result.error("-5", "Recorder is not initialized", null)
    }

    // Handles the "pause" method call to pause recording
    private fun handlePause(result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            recorder?.pause(result) ?: result.error("-5", "Recorder is not initialized", null)
        } else {
            result.error("-6", "Pause is not supported on this Android version", null)
        }
    }

    // Handles the "resume" method call to resume recording
    private fun handleResume(result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            recorder?.resume(result) ?: result.error("-5", "Recorder is not initialized", null)
        } else {
            result.error("-6", "Resume is not supported on this Android version", null)
        }
    }

    // Handles the "isPaused" method call to check if recording is paused
    private fun handleIsPaused(result: MethodChannel.Result) {
        recorder?.isPaused(result) ?: result.error("-5", "Recorder is not initialized", null)
    }

    // Handles the "isRecording" method call to check if recording is in progress
    private fun handleIsRecording(result: MethodChannel.Result) {
        recorder?.isRecording(result) ?: result.error("-5", "Recorder is not initialized", null)
    }

    // Handles the "getAmplitude" method call to get the current amplitude
    private fun handleGetAmplitude(result: MethodChannel.Result) {
        recorder?.getAmplitude(result) ?: result.error("-5", "Recorder is not initialized", null)
    }

    // Handles the "dispose" method call to release resources
    private fun handleDispose(result: MethodChannel.Result) {
        recorder?.close() ?: result.error("-5", "Recorder is not initialized", null)
        result.success(null)
    }

    // Handles the "hasPermission" method call to check for audio recording permission
    private fun handleHasPermission(result: MethodChannel.Result) {
        if (!isPermissionGranted) {
            pendingPermResult = result
            askForPermission()
        } else {
            result.success(true)
        }
    }

    // Checks if the audio recording permission is granted
    private val isPermissionGranted: Boolean
        get() {
            val activity = this.activity ?: return false
            val result = ActivityCompat.checkSelfPermission(activity, Manifest.permission.RECORD_AUDIO)
            return result == PackageManager.PERMISSION_GRANTED
        }

    // Requests audio recording permission
    private fun askForPermission() {
        val activity = this.activity
        if (activity == null) {
            pendingPermResult?.error("-1", "Activity is null", null)
            pendingPermResult = null
            return
        }
        ActivityCompat.requestPermissions(
            activity,
            arrayOf(Manifest.permission.RECORD_AUDIO),
            RECORD_AUDIO_REQUEST_CODE
        )
    }

    // Handles the result of a permission request
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray): Boolean {
        if (requestCode == RECORD_AUDIO_REQUEST_CODE) {
            if (pendingPermResult != null) {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    pendingPermResult!!.success(true)
                } else {
                    pendingPermResult!!.success(false)
                }
                pendingPermResult = null
                return true
            }
        }
        return false
    }

    // Returns the file extension based on the encoder type
    private fun getFileExtension(encoder: Int): String {
        return when (encoder) {
            6 -> ".wav" // WAV format
            3, 4 -> ".amr" // AMR format
            5 -> ".opus" // Opus format
            else -> ".m4a" // Default to AAC format
        }
    }
}